azkaban-common/src/main/java/azkaban/executor/dispatcher/CandidateComparator.java

```diff
@@ -0,0 +1,187 @@
+      if (item != null){
+        totalWeight += item.getWeight();
+      }
+    }
+
+    return totalWeight;
+  }
+
+  /** function to actually calculate the scores for the two objects that are being compared.
+   *  the comparison follows the following logic -
+   *  1. if both objects are equal return 0 score for both.
+   *  2. if one side is null, the other side gets all the score.
+   *  3. if both sides are non-null value, both values will be passed to all the registered FactorComparators
+   *     each factor comparator will generate a result based off it sole logic the weight of the comparator will be
+   *     added to the wining side, if equal, no value will be added to either side.
+   *  4. final result will be returned in a Pair container.
+   *
+   * */
+  public Pair<Integer,Integer> getReult(T object1, T object2){
+    logger.info(String.format("start comparing '%s' with '%s',  total weight = %s ",
+        object1 == null ? "(null)" : object1.toString(),
+        object2 == null ? "(null)" : object2.toString(),
+        this.getTotalWeight()));
+
+    // short cut if object equals.
+    if (object1 ==  object2){
+      logger.info("Result : 0 vs 0 (equal)");
+      return new Pair<Integer,Integer>(0,0);
+    }
+
+    // left side is null.
+    if (object1 == null){
+      logger.info(String.format("Result : 0 vs %s (left is null)",this.getTotalWeight()));
+      return new Pair<Integer, Integer>(0,this.getTotalWeight());
+    }
+
+    // right side is null.
+    if (object2 == null){
+      logger.info(String.format("Result : %s vs 0 (right is null)",this.getTotalWeight()));
+      return new Pair<Integer, Integer>(this.getTotalWeight(),0);
+    }
+
+    // both side is not null,put them thru the full loop
+    int result1 = 0 ;
+    int result2 = 0 ;
+    Collection<FactorComparator<T>> comparatorList = this.factorComparatorList.values();
+    Iterator<FactorComparator<T>> mapItr = comparatorList.iterator();
+    while (mapItr.hasNext()){
+      FactorComparator<T> comparator = (FactorComparator<T>) mapItr.next();
+      int result = comparator.compare(object1, object2);
+      result1  = result1 + (result > 0 ? comparator.getWeight() : 0);
+      result2  = result2 + (result < 0 ? comparator.getWeight() : 0);
+      logger.info(String.format("[Factor: %s] compare result : %s (current score %s vs %s)",
+          comparator.getFactorName(), result, result1, result2));
+    }
+
+    // in case of same score, user differentiator to stabilize the result.
+    if (result1 == result2){
+      boolean result = this.differentiate(object1, object2);
+      logger.info("[Differentiator] differentiator chose " + (result?  object1.toString(): object2.toString()));
+      if (result) result1++; else result2++;
+    }
+
+    logger.info(String.format("Result : %s vs %s ",result1,result2));
+    return new Pair<Integer,Integer>(result1,result2);
+  }
+
+  @Override
+  public int compare(T o1, T o2) {
```

Comment:
Yes, I am looking for "for this specific Flow tell me whether executor A or executor B suits better ?" part. I think it makes sense to add this in interface (even if we are not use it), to give flexibility to extend and use other features derived out of ExecutableFlow and Executor combination. 

For your other concern about end user manipulating properties to prefer some executor. If user is able to directly manipulate then there is a flaw in new admin level implementation which derived new features, on the other hand some properties will obviously lead to prefer some executors and that can be desirable in certain situation e.g. we may prefer to run spark jobs on current high memory instance but will also work with other instance due to other issues like memory, In an multi-cluster executors we may allow users to show affinity to certain cluster and help them avoid distcp, etc These types of cases cannot be accommodated by filters and need to have ExecutableFlow or a derived feature vector.

Yes, template solves that thing in this scenario.